node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1454 @@
+
+			for id in ids.iter() {
+				validator_indices.set(*id as usize, true);
+			}
+
+			let backed = BackedCandidate {
+				candidate,
+				validity_votes,
+				validator_indices,
+			};
+
+			res.push(NewBackedCandidate(backed.clone()));
+		}
+
+		res
+	}
+
+	async fn issue_new_misbehaviors(&mut self) -> Result<(), Error> {
+		let mut reports = Vec::new();
+
+		for (k, v) in self.table.get_misbehavior().iter() {
+			if !self.reported_misbehavior_for.contains(k) {
+				self.reported_misbehavior_for.insert(*k);
+
+				let f = FromTableMisbehavior {
+					id: *k,
+					report: v.clone(),
+					signing_context: self.table_context.signing_context.clone(),
+					key: self.table_context.validators[*k as usize].clone(),
+				};
+
+				if let Ok(report) = MisbehaviorReport::try_from(f) {
+					let message = ProvisionerMessage::ProvisionableData(
+						ProvisionableData::MisbehaviorReport(self.parent, report)
+					);
+
+					reports.push(message);
+				}
+			}
+		}
+
+		for report in reports.drain(..) {
+			self.send_to_provisioner(report).await?
+		}
+
+		Ok(())
+	}
+
+	async fn import_statement(&mut self, statement: SignedFullStatement) -> Result<(), Error> {
+		let idx = statement.validator_index() as usize;
+
+		if self.table_context.validators.len() > idx {
+			match statement.check_signature(
+				&self.table_context.signing_context,
+				&self.table_context.validators[idx],
+			) {
+				Ok(()) => {
+					let stmt = primitive_statement_to_table(&statement);
+
+					self.table.import_statement(&self.table_context, stmt);
+
+					self.issue_new_misbehaviors().await?;
+				}
+				Err(()) => {
+					return Err(Error::ValidationFailed(ValidationFailed));
+				}
+			}
+		};
+
+		Ok(())
+	}
+
+	async fn process_msg(&mut self, msg: CandidateBackingMessage) -> Result<(), Error> {
+		match msg {
+			CandidateBackingMessage::Second(_, candidate, pov) => {
+				// If the message is a `CandidateBackingMessage::Second`, sign and dispatch a
+				// Seconded statement only if we have not seconded any other candidate and
+				// have not signed a Valid statement for the requested candidate.
+				match self.seconded {
+					// This job has not seconded a candidate yet.
+					None => {
+						let candidate_hash = candidate.hash();
+
+						if !self.issued_validity.contains(&candidate_hash) {
+							if let Ok(ValidationResult::Valid) = self.validate_and_second(
+								candidate,
+								pov,
+							).await {
+								self.seconded = Some(candidate_hash);
+							}
+						}
+					}
+					// This job has already seconded a candidate.
+					Some(_) => {}
+				}
+			}
+			CandidateBackingMessage::Statement(_, statement) => {
+				self.maybe_validate_and_import(statement).await?;
+			}
+			CandidateBackingMessage::GetBackedCandidates(_, tx) => {
+				let backed = self.get_backed();
+
+				tx.send(backed).map_err(|_| oneshot::Canceled)?;
+			}
+		}
+
+		Ok(())
+	}
+
+	async fn maybe_validate_and_import(
```

Comment:
I'm bikeshedding, but I don't feel this name captures the fact that this does kick off validation work, if necessary.

Reply:
The main reason I bring it up is that it might make sense to refactor this into two functions: one for importing external statements and one for kicking off validation work.

Actually, we already have a function for checking signature and importing the statement. It's called `self.`import_statement. The inner table returns an `Option<Summary>` that includes both the digest and the group ID.

So a simple `kick_off_validation_work(Summary)` might be enough, because the table also has a `get_candidate` function. Then we can kick off validation work for any candidate in our validation group. That would make most of this function redundant.